### DISCUSSION TITLE
A Turorial for the: Tutorial For Nervous Beginners (Win10)

### DISCUSSION THREAD

**Discussion by dbuchacher** [coqui-ai/TTS#1332] [Show and tell]
# From Zero to something using windows 10

I came across this project and thought it was cool that I could train a voice to be used as a TTS.  Just to get the example training to work has been a struggle to say the least.  I have just finished getting the example to work so I wanted to share what I did in hopes somebody else finds it useful.

### install Linux sub system

>press windows key
>type `power shell`
>right-click on the power shell icon
>select `run as administrator`

```powershell
wsl --install # enter this command to install the linux sub system
```

>reboot pc to finish install

> once finished a console window will prompt you to enter a new username and password

```shell
sudo apt update && sudo apt upgrade # then install updates
```

```shell
sudo apt install python3-pip # install pip
```

---

### following https://tts.readthedocs.io/en/latest/tutorial_for_nervous_beginners.html

```shell
git clone https://github.com/coqui-ai/TTS
cd TTS
sudo pip install -e .
```

```bash
python3 -c 'from TTS.utils.downloaders import download_ljspeech; download_ljspeech("../recipes/ljspeech/");'
```

```bash
nano train.py # create new file
```

>copy paste the below python script into the nano editor screen
>ctrl + s to save script
>ctrl + x to exit nano 

```python
import os

# Trainer: Where the ✨️ happens.
# TrainingArgs: Defines the set of arguments of the Trainer.
from TTS.trainer import Trainer, TrainingArgs

# GlowTTSConfig: all model related values for training, validating and testing.
from TTS.tts.configs.glow_tts_config import GlowTTSConfig

# BaseDatasetConfig: defines name, formatter and path of the dataset.
from TTS.tts.configs.shared_configs import BaseDatasetConfig
from TTS.tts.datasets import load_tts_samples
from TTS.tts.models.glow_tts import GlowTTS
from TTS.utils.audio import AudioProcessor

# we use the same path as this script as our training folder.
output_path = os.path.dirname(os.path.abspath(__file__))

# DEFINE DATASET CONFIG
# Set LJSpeech as our target dataset and define its path.
# You can also use a simple Dict to define the dataset and pass it to your custom formatter.
dataset_config = BaseDatasetConfig(
    name="ljspeech", meta_file_train="metadata.csv", path=os.path.join(output_path, "../recipes/ljspeech/LJSpeech-1.1/")
)

# INITIALIZE THE TRAINING CONFIGURATION
# Configure the model. Every config class inherits the BaseTTSConfig.
config = GlowTTSConfig(
    batch_size=32,
    eval_batch_size=16,
    num_loader_workers=4,
    num_eval_loader_workers=4,
    run_eval=True,
    test_delay_epochs=-1,
    epochs=1000,
    text_cleaner="phoneme_cleaners",
    use_phonemes=True,
    phoneme_language="en-us",
    phoneme_cache_path=os.path.join(output_path, "phoneme_cache"),
    print_step=25,
    print_eval=False,
    mixed_precision=True,
    output_path=output_path,
    datasets=[dataset_config],
)

# INITIALIZE THE AUDIO PROCESSOR
# Audio processor is used for feature extraction and audio I/O.
# It mainly serves to the dataloader and the training loggers.
ap = AudioProcessor(**config.audio.to_dict())

# LOAD DATA SAMPLES
# Each sample is a list of ```[text, audio_file_path, speaker_name]```
# You can define your custom sample loader returning the list of samples.
# Or define your custom formatter and pass it to the `load_tts_samples`.
# Check `TTS.tts.datasets.load_tts_samples` for more details.
train_samples, eval_samples = load_tts_samples(dataset_config, eval_split=True)

# INITIALIZE THE MODEL
# Models take a config object and a speaker manager as input
# Config defines the details of the model like the number of layers, the size of the embedding, etc.
# Speaker manager is used by multi-speaker models.
model = GlowTTS(config, speaker_manager=None)

# INITIALIZE THE TRAINER
# Trainer provides a generic API to train all the 🐸TTS models with all its perks like mixed-precision training,
# distributed training, etc.
trainer = Trainer(
    TrainingArgs(),
    config,
    output_path,
    model=model,
    train_samples=train_samples,
    eval_samples=eval_samples,
    training_assets={"audio_processor": ap},  # assets are objetcs used by the models but not class members.
)

# AND... 3,2,1... 🚀
trainer.fit()
```

```shell
CUDA_VISIBLE_DEVICES=0 python3 train.py # start training
```

> I was seeing a successful train in the console at this point.

----

### NOTES

1. 
>`pip install -e .` failed for me when ran it the 1st time so I tried
>`sudo pip install -e .` which was completed successfully afterwards 

2. 
> `../LJSpeech-1`
> was changed to
> `../recipes/ljspeech/LJSpeech-1.1/`
> in the train.py script

3. 
> `python` command didn't work.
> hand to use `python3`

4. 
> I personally had no issue with it starting and using my video card
> 
> BUT the day prior I had used https://stackoverflow.com/questions/66726331/how-can-i-run-mozilla-tts-coqui-tts-training-with-cuda-on-a-windows-system
>
> to install CUDA and eSpeak (so if you are having issues with video card this might be why)
>
> this part
> [Download](https://developer.nvidia.com/cuda-10.1-download-archive-base) and install CUDA Toolkit 10.1 (_not_ 11.0+).
> 
> [Download](https://developer.nvidia.com/rdp/cudnn-archive) "cuDNN v7.6.5 (November 5th, 2019), for CUDA 10.1" (_not_ cuDNN v8+), extract it, and then copy what's inside the `cuda` folder into `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1`.
> 
> [Download](https://github.com/espeak-ng/espeak-ng/releases) the latest 64-bit version of eSpeak NG (no version constraints :-) ).

Good Luck

**Comment by Maxello1** [comment]
![image](https://user-images.githubusercontent.com/98433747/183753536-637c3287-703b-498b-b906-d15320641e42.png)
what's up with this error? I have cuda and cuDNN installed and also espeak NG, do you know what could cause this?

**Reply by RavingSmurfGB** [reply]
yeah - you have to use the new isntructions
https://tts.readthedocs.io/en/latest/tutorial_for_nervous_beginners.html

train.py imports have changed

**Comment by KyuubiYoru** [comment]
"install CUDA Toolkit 10.1" doesn't work with 30xx cards, had to install cuda 11

**Comment by thorstenMueller** [comment]
Just in case you want to train a TTS model on Windows without WSL. I've created a step by step tutorial on that.
https://youtu.be/bJjzSo_fOS8

**Comment by Cooperdevi** [comment]
I get an error:
`PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:/Users/hakan/Documents/recipes/run-July-28-2023_10+11PM-0000000\\trainer_0_log.txt'   `     
when running train.py.
Any idea?

**Reply by kylerifqi** [reply]
did you ever find a fix?

**Reply by KyuubiYoru** [reply]
You have to use "freeze_support()"

```
from multiprocessing import freeze_support

if __name__ == "__main__":
    freeze_support()
    trainer.fit()
```
